Fix cross compilation of LLVM-3.9

expipiplus1 opened the pull request:
~~Although LLVM now cross compiles successfully the executable do not run, the
`RUNPATH` of the resultant executables in incorrect, it does not contain the
LLVM lib output.~~

expipiplus1 committed:
clang: Use cmake from buildPackages

expipiplus1 committed:
llvm-3.9: Fix cross compilation

It's now possible to cross compile llvm:

`nix-build -E '(import ./. { crossSystem = import ./platform.nix; }).pkgs.llvm'`

expipiplus1 committed:
groff: use buildPackages to fix cross compile

Ericson2314 commented on pkgs/development/compilers/llvm/3.9/llvm.nix:
```diff
@@ -39,7 +42,13 @@ in stdenv.mkDerivation rec {
 
-  buildInputs = [ perl groff cmake libxml2 python libffi ]
+  buildInputs = [
+    buildPackages.perl
```
Can these be `nativeBuildInputs`? might not even cause rebuild for native due to some normalizing in stdenv.

Ericson2314 commented on pkgs/development/compilers/llvm/3.9/clang/default.nix:
```diff
@@ -13,7 +13,7 @@ let
     '';
 
-    buildInputs = [ cmake libedit libxml2 llvm python ];
+    buildInputs = [ buildPackages.cmake libedit libxml2 llvm python ];
```
`cmake` should also be a `nativeBuildInput`. Note that `buildPackages.` is not necessary in this case because `mkDerivation` will magically fish it out.

Ericson2314 commented on pkgs/development/compilers/llvm/3.9/llvm.nix:
```diff
@@ -88,6 +97,9 @@ in stdenv.mkDerivation rec {
     "-DLLVM_ENABLE_LIBCXX=ON"
     "-DCAN_TARGET_i386=false"
+  ] ++ stdenv.lib.optionals (buildPlatform != hostPlatform) [
+    "-DCMAKE_CROSSCOMPILING=True"
```
Hmm, this is a general cmake thing? This could be generalized to be so added for all derivations with `cmakeFlags` I suppose.

expipiplus1 replied:
Looks like it https://cmake.org/cmake/help/v3.0/variable/CMAKE_CROSSCOMPILING.html

expipiplus1 committed:
LLVM-3.9: Fix RPATH in cross compile builds

This error was cause by multiple-outputs.sh not setting
NIX_CROSS_LDFLAGS

expipiplus1 committed:
LLVM-3.9: Use nativeBuildInputs instead of buildPackages

Ericson2314 commented on pkgs/development/compilers/llvm/3.9/llvm.nix:
```diff
@@ -39,7 +42,16 @@ in stdenv.mkDerivation rec {
+  buildInputs = [
+    groff
+    libxml2
+    libffi ]
```
Not to be a big nit, but `]` on the line below before `++` I think would be more idiomatic.

expipiplus1 committed:
LLVM-3.9: formatting

expipiplus1 committed:
LLVM-3.9: disable shared libraries in cross builds

The current cc-wrapper script seems to have trouble setting the rpath
correctly. Hopefully #25047 will fix this.

7c6f434c commented:
@expipiplus1 do you consider this ready to merge? Given LLVM 3.9 is what Mesa currently uses, I would target this to `staging`

Ericson2314 commented:
It's not entirely clear to me this is/need-be a hash-breaking change given the way *inputs are normalized.

It does look good to me, so definitely want to merge somewhere.

7c6f434c commented:
I think the order of inputs needs to change (because the current situation means interleaving of native and target inputs), and that's a hash change.

expipiplus1 commented:
@7c6f434c Yeah, I think it's good to go. I'd like to get to the bottom of why the shared library support doesn't work at some time.

expipiplus1 commented:
I'm in no hurry though!

7c6f434c commented:
Well, I guess when you want it merged you want to retarget it to the `staging` branch

expipiplus1 commented:
done!
